amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,594 @@
+from abc import abstractproperty
+from fractions import Fraction
+import re
+
+from ..hdl import *
+from ..lib.cdc import ResetSynchronizer
+from ..build import *
+
+# Acknowledgments:
+#   Parts of this file originate from https://github.com/tcjie/Gowin
+
+
+__all__ = ["GowinPlatform"]
+
+
+class GowinPlatform(TemplatedPlatform):
+    """
+    .. rubric:: Apicula toolchain
+
+    Required tools:
+        * ``yosys``
+        * ``nextpnr-gowin``
+        * ``gowin_pack``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_APICULA``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+
+    .. rubric:: Gowin toolchain
+
+    Required tools:
+        * ``gw_sh``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_GOWIN``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+    """
+
+    toolchain = None # selected when creating platform
+
+    part   = abstractproperty()
+    family = abstractproperty()
+
+    def parse_part(self):
+        # These regular expressions match all >900 parts of Gowin device_info.csv
+        reg_series    = r"(GW[12]{1}[AN]{1}[EFNRSZ]{0,3})-"
+        reg_voltage   = r"(ZV|EV|LV|LX|UV|UX)"
+        reg_size      = r"(1|2|4|9|18|55)"
+        reg_subseries = r"(?:(B|C|S|X|P5)?)"
+        reg_package   = r"((?:PG|UG|EQ|LQ|MG|M|QN|CS|FN)(?:\d+)(?:P?)(?:A|E|M|CF|C|D|G|H|F|S|T|U|X)?)"
+        reg_speed     = r"((?:C\d{1}/I\d{1})|ES|A\d{1}|I\d{1})"
+
+        match = re.match(reg_series+reg_voltage+reg_size+reg_subseries+reg_package+reg_speed+"$",
+                         self.part)
+        if not match:
+            raise ValueError("Supplied part name is invalid")
+
+        self.series    = match.group(1)
+        self.voltage   = match.group(2)
+        self.size      = match.group(3)
+        self.subseries = match.group(4) or ""
+        self.package   = match.group(5)
+        self.speed     = match.group(6)
+
+        match = re.match(reg_series+reg_size+reg_subseries+"$", self.family)
```

Comment:
Hold on. This matches series, size, and subseries, *all of which are also matched above*. So what information does `family` actually add?

Reply:
It's... Complicated. Iirc device names don't actually uniquely define the actual device. I think I have an email from their support that's supposed to explain it but honestly it just doesn't make any sense.

Reply:
Have you looked at the code here?

Reply:
It might be related to *revisions* of the chips, as shown on the package marking, as opposed to the datasheets:

https://github.com/brilliantlabsAR/docs/blob/e075bbe5784e6a61d23e10cbb86e26fe74a1831b/monocle/monocle.md#fpga-chip-revisions

So while a full part order name might be `GW1N-LV9 MG100C6/I5` the device family name *as shown in GOWIN EDA* software tool might be different according to... the extra chip marking information "xxxx**C**A0N" also shown here: `GW1N-9C`

Many thanks to @siliconwitch who found out about it.

And as you can see, this would also have a different Device ID.

Reply:
That is nuts.

Reply:
@josuah That is exactly why I'm doing the double-matching and why one also has to supply `family` to uniquely identify the part.

Reply:
Just take a look at the contents of `<path-to-gowin-IDE>/share/device/device_info.csv`.
Run this:

`awk -F"," '{print $2}' device_info.csv  | sort | uniq -c`

and you'll see that the part name in column 2 is *not unique* enough.

Reply:
And the device family lacks the package information used for the pin mapping.

I suppose they would have preferred to avoid this split themselves.
Maybe a difference in the process with the same masks.
Maybe a fix that lead to more consequences than intended.

Reply:
@bl0x But the RE that matches `family` is a subset of the RE that matches `part`... what am I missing?

Reply:
@whitequark Nothing! Because part can also contain all information (and oftentimes does), but not always. We could use the same RE to also match `family`, but that would be too general.